File: 347.py
# Original approach: use counter and sort items
    # this is easy but unnecessary b/c we only need the first k more frequent elements
# Second approach: use heap
    # cba'd implementing because heapq is annoying with alternate keys
# Best approach: bucket sort
    # Create array where the index equals the count,
        # and the values are the elements having that particular count
    # Using this array, can just iterate backwards until we observe k elements
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        c = Counter(nums)
        n = len(nums)
        bucket = [[] for i in range(n+1)]

        # Populate bucket
        for i in c.items():
            element, count = i
            bucket[count].append(element)

        # Iterate backwards through bucket and add k elements
        output = []
        idx = n
        while len(output) < k:
            if bucket[idx]: output.extend(bucket[idx])
            idx -= 1

        return output

File: C++/347.cpp
/*
- Easy solution: sort counter or use heap
- Better solution: use bucket sort
    - As usual, set up a counter
    - Then, make a vector where each index contains a vector of elements that happen
    - that amount of times
*/
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;
        for(int i = 0; i < nums.size(); i++){
            map[nums[i]]++;
        }

        vector<vector<int>> buckets(nums.size() + 1);
        for(auto it = map.begin(); it != map.end(); it++){
            buckets[it -> second].push_back(it -> first);
        }

        vector<int> result;
        for(int i = nums.size(); i >= 0; i--){
            if(result.size() >= k) break;
            if(!buckets[i].empty()){
                result.insert(result.end(), buckets[i].begin(), buckets[i].end());
            }
        }
        return result;
    }
};